src/Peachpie.CodeAnalysis/FlowAnalysis/AnalysisVisitor.cs

```diff
@@ -327,23 +332,32 @@ public override void VisitCFGExitBlock(ExitBlock x)
             VisitCFGBlock(x);
 
             // TODO: EdgeToCallers:
-            EnqueueSubscribers(x);
+            PingSubscribers(x);
         }
 
-        protected void EnqueueSubscribers(ExitBlock exit)
+        protected void PingSubscribers(ExitBlock exit)
         {
             if (exit != null)
             {
+                bool wasNotAnalysed = false;
+                if (_state.Routine?.IsReturnAnalysed == false)
+                {
+                    wasNotAnalysed = true;
+                    _state.Routine.IsReturnAnalysed = true;
+                }
+
+                // Ping the subscribers either if the return type has changed or
+                // it is the first time the analysis reached the routine exit
                 var rtype = _state.GetReturnType();
-                if (rtype != exit._lastReturnTypeMask)
+                if (rtype != exit._lastReturnTypeMask || wasNotAnalysed)
                 {
                     exit._lastReturnTypeMask = rtype;
                     var subscribers = exit.Subscribers;
                     if (subscribers.Count != 0)
                     {
                         lock (subscribers)
                         {
-                            subscribers.ForEach(_worklist.Enqueue);
+                            subscribers.ForEach(subscriber => _worklist.PingReturnUpdate(exit, subscriber));
```

Comment:
this creates an ugly closure - change it to simple `foreach (var s in subscribers) _worlist.Ping...;` pls